### PR TITLE
Support for AtomicCmpXchgInst and AtomicRMWInst

### PR DESCRIPTION
**Pull request by caballa**

### CODE REVIEW

File: src/DsaLocal.cc [hunk 1]
```diff
@@ -533,6 +533,62 @@ void IntraBlockBuilder::visitLoadInst(LoadInst &LI) {
   }
 }
 
+/// OldVal := *Ptr
+/// if OldVal == Cmp then
+///    *Ptr := New
+/// return OldVal
```
Comment (kuhar) [inline src/DsaLocal.cc]: Should this return the other value, i.e., `if *Ptr == New then Old else New`?

File: src/DsaLocal.cc [hunk 2]
```diff
@@ -533,6 +533,62 @@ void IntraBlockBuilder::visitLoadInst(LoadInst &LI) {
...
+  
+  Cell PtrC = valueCell(*Ptr);
+  assert(!PtrC.isNull());
+    
+  PtrC.setModified();
+  PtrC.setRead();
+  PtrC.growSize(0, I.getType());
+  PtrC.addAccessedType(0, I.getType());
+
+  if (!isSkip(I)) {
```
Comment (kuhar) [inline src/DsaLocal.cc]: Isn't `I`'s type a strutcture type `{New->getType(), i1}`?

```
  %13 = cmpxchg weak i64* %8, i64 %11, i64 %12 monotonic monotonic
  %14 = extractvalue { i64, i1 } %13, 0
  %15 = extractvalue { i64, i1 } %13, 1
```

This seems to skip over all the time, no?

File: src/DsaLocal.cc (hunk 1 above)
Comment (caballa) [reply src/DsaLocal.cc]: This is the fixed semantics:
```
/// OldVal := *Ptr
/// Success := false
/// if OldVal == Cmp then
///    Success := true
///    *Ptr := New
/// return {OldVal, Success}
```

File: src/DsaLocal.cc (hunk 2 above)
Comment (caballa) [reply src/DsaLocal.cc]: Fixed

**Comment by shaobo-he** [comment]
Hey guys,

Any updates on this PR?

Thanks,
Shaobo

**Comment by agurfinkel** [comment]
@shaobo-he merged into llvm5 branch. will sync with llvm8 next

**Comment by shaobo-he** [comment]
> @shaobo-he merged into llvm5 branch. will sync with llvm8 next

Thank you very much.

**Pull request merged**